server/src/main/scala/com/cloudera/livy/server/ApiVersioningSupport.scala

```
@@ -0,0 +1,101 @@
+
+import org.scalatra.{NotAcceptable, ScalatraBase}
+
+/**
+ * Livy's servlets can mix-in this trait to get API version support.
+ *
+ * Example: {{{
+ * import ApiVersions._
+ * class FooServlet
+ *   ...
+ *   with ApiVersioningSupport
+ *   ...
+ * {
+ *   get("/test") {
+ *     ...
+ *   }
+ *   get("/test", getApiVersion() <= v0_2) {
+ *     ...
+ *   }
+ *   get("/test", getApiVersion() <= v0_1) {
+ *     ...
+ *   }
+ * }
+ * }}}
+ */
+trait ApiVersioningSupport extends AbstractApiVersioningSupport {
+  this: ScalatraBase =>
+  // Link the abstract trait to Livy's version enum.
+  override val apiVersions = ApiVersions
+  override type ApiVersionType = ApiVersions.Value
+}
+
+trait AbstractApiVersioningSupport {
+  this: ScalatraBase =>
+  protected val apiVersions: Enumeration
+  protected type ApiVersionType
+
+  /**
+   * Before proceeding with routing, validate the specified API version in the request.
+   * If validation passes, cache the parsed API version as a per-request attribute.
+   */
+  before() {
+    try {
+      request(AbstractApiVersioningSupport.ApiVersionKey) = parseApiVersion(request)
+    } catch {
+      case e: Exception =>
+        halt(NotAcceptable(e.getMessage))
+    }
+  }
+
+  /**
+   * @return The specified API version in the request.
+   */
+  def getApiVersion(): ApiVersionType = {
+    request(AbstractApiVersioningSupport.ApiVersionKey).asInstanceOf[ApiVersionType]
+  }
+
+  private def parseApiVersion(request: HttpServletRequest): ApiVersionType = {
+    // Get every character after "application/vnd.livy.v" until hitting a + sign.
+    val acceptHeaderRegex = """application/vnd\.livy\.v([^\+]*).*""".r
+    request.getHeader("Accept") match {
+      case acceptHeader @ acceptHeaderRegex(apiVersion) =>
+        try {
+          apiVersions.withName(apiVersion).asInstanceOf[ApiVersionType]
+        } catch {
+          case e: NoSuchElementException =>
+            throw new Exception(s"Unknown API version: $acceptHeader", e)
```

I guess what I don't understand is why you want that. There's a single call site for that function. The call site, ignoring exception handling, is a single line of code. That screams "inline me!".

If you had unit tests that called `parseApiVersion` directly, I'd sort of understand the desire. But you don't, so your desire to separate concerns just complicates the code for no reason. The whole method could be inlined in the `before()` handler and nothing would change, except you'd have less code, and you wouldn't have this weird contract "let me throw a very generic exception just because the caller is not interested in the exception, but in its message".